{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Summary\n",
    "Show how to retrieve list of genes for specified organisms\n",
    "\n",
    "## Workflow\n",
    "\n",
    "For a given assembly `accession.version`\n",
    "* Get a ref-seq id for assembly\n",
    "* Get a nuccore id for refseq-id\n",
    "* List all gene-ids for a assembly by its nuccoreid\n",
    "* For each gene\n",
    "    * Get gene info and display it   \n",
    "    * Get some pubmed relateed information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "code_folding": [],
    "hide_input": false
   },
   "outputs": [],
   "source": [
    "import requests\n",
    "import xml.etree.ElementTree as xml\n",
    "from IPython.display import HTML, display\n",
    "from Bio import Entrez\n",
    "\n",
    "#email Set the Entrez email parameter (default is not set).\n",
    "Entrez.email = \"great_team@hackathon.ncbi.org\"\n",
    "\n",
    "#tool Set the Entrez tool parameter (default is biopython).\n",
    "Entrez.tool = \"hackathon_examples\"\n",
    "\n",
    "def get_nuccore_id(uid):\n",
    "    \"\"\"\n",
    "    Get nuccore id by its refseq id.\n",
    "    \"\"\"\n",
    "    url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi'\n",
    "    params = {'dbfrom': 'assembly', 'db':'nuccore', 'retmode':'json', 'id': uid}\n",
    "    r = requests.get(url, params=params)\n",
    "    if r.status_code != 200:\n",
    "        raise Exception(\"Cant nuccore id for specified assembly\")\n",
    "    doc = r.json()['linksets'][0]['linksetdbs']\n",
    "    for link in doc:\n",
    "        if link['linkname'] == \"assembly_nuccore_refseq\":\n",
    "            return int(link['links'][0])\n",
    "    else:\n",
    "        return int(doc[0]['links'][0])\n",
    "\n",
    "    \n",
    "def list_gene_ids(nuc_id):\n",
    "    \"\"\"\n",
    "    List genes for specified organism nuccore id.\n",
    "    \"\"\"\n",
    "    url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi'\n",
    "    params = {'dbfrom':'nuccore', 'db': 'gene', 'retmode': 'json', 'id':nuc_id}\n",
    "    r = requests.get(url, params=params)\n",
    "    if r.status_code != 200:\n",
    "        raise Exception(\"Cant find annotation\")\n",
    "    return r.json()['linksets'][0]['linksetdbs'][0]['links']\n",
    "\n",
    "\n",
    "def list_gene_ids_alt(refseq_id):\n",
    "    \"\"\"\n",
    "    List genes for specified organism nuccore id.\n",
    "    \"\"\"\n",
    "    url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi'\n",
    "    params = {'dbfrom':'assembly', 'db': 'genome', 'retmode': 'json', 'id':refseq_id}\n",
    "    r = requests.get(url, params=params)\n",
    "    if r.status_code != 200:\n",
    "        raise Exception(\"Cant find refseq\")\n",
    "    genome_id = r.json()['linksets'][0]['linksetdbs'][0]['links'][0]\n",
    "    \n",
    "    \n",
    "    url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi'\n",
    "    params = {'dbfrom':'genome', 'db': 'gene', 'retmode': 'json', 'id':genome_id}\n",
    "    r = requests.get(url, params=params)\n",
    "    if r.status_code != 200:\n",
    "        raise Exception(\"Cant find annotation\")\n",
    "   \n",
    "    return r.json()['linksets'][0]['linksetdbs'][0]['links']\n",
    "\n",
    "\n",
    "def get_asm_details(accession):\n",
    "    \"\"\"\n",
    "    Get ref_seq id and organism name for specified assembly.\n",
    "    \"\"\"\n",
    "    params = {'release':accession}\n",
    "    r = requests.get(\n",
    "        'https://www.ncbi.nlm.nih.gov/projects/r_gencoll/asm4portal/gc_assembly_svc.cgi',\n",
    "        params=params)\n",
    "    if r.status_code != 200:\n",
    "        raise Exception(\"Cant find assembly\")\n",
    "    doc = r.content.decode('utf-8')\n",
    "    root = xml.fromstring(doc)\n",
    "    refseq_id = int(root.attrib['uid'])\n",
    "    header = root.find('header')\n",
    "    organism_name = header.find('species-organism').text\n",
    "    return refseq_id, organism_name\n",
    "\n",
    "\n",
    "def build_gene_table(genes, max_show, organism_name):\n",
    "    \"\"\"\n",
    "    Build html table listing gene names, proteins and publication numbers.\n",
    "    \"\"\"\n",
    "\n",
    "    html = \"<table>\"\n",
    "\n",
    "    html += \"<tr>\"       \n",
    "    html += \"<th> Gene Tag </th>\"\n",
    "    html += \"<th> Gene Locus </th>\"\n",
    "    html += \"<th> Protein </th>\"\n",
    "    html += \"<th> Protein sequence </th>\"\n",
    "    html += \"<th> Number of Pubmed publications </th>\"\n",
    "    html += \"</tr>\"\n",
    "    \n",
    "    for gene in genes[0:max_show]:\n",
    "        handle = Entrez.efetch(db=\"gene\", id=gene, retmode=\"xml\")\n",
    "        doc = handle.read()\n",
    "        root = xml.fromstring(doc)\n",
    "        gene_node = root.find('Entrezgene')\n",
    "        gene_ref = gene_node.find('Entrezgene_gene').find('Gene-ref')\n",
    "        \n",
    "        locus_at = gene_ref.find('Gene-ref_locus')\n",
    "        locus_tag = gene_ref.find('Gene-ref_locus-tag')\n",
    "\n",
    "        terms = []\n",
    "        locus = \"n/a\"\n",
    "        tag = \"n/a\"\n",
    "\n",
    "        try:\n",
    "            tag = locus_tag.text\n",
    "            terms.append(tag)\n",
    "        except:\n",
    "            pass\n",
    "        \n",
    "        try:\n",
    "            locus = locus_at.text\n",
    "            terms.append(locus)\n",
    "        except:\n",
    "            pass\n",
    "        \n",
    "        prot_node = gene_node.find('Entrezgene_prot')\n",
    "        \n",
    "        prot = \"n/a\"\n",
    "        \n",
    "        if prot_node:\n",
    "            prot = prot_node.find('Prot-ref').find('Prot-ref_name').find('Prot-ref_name_E').text\n",
    "            terms.append(prot)\n",
    "            \n",
    "            \n",
    "        loc_node = gene_node.find('Entrezgene_locus')\n",
    "        com = loc_node.find('Gene-commentary')\n",
    "        prod = com.find('Gene-commentary_products')\n",
    "        prod_seq = \"n/a\"\n",
    "        prod_ver = \"n/a\"\n",
    "        if prod:\n",
    "            prod_com = prod.find('Gene-commentary')\n",
    "            if prod_com:\n",
    "                prod_seq = prod_com.find('Gene-commentary_accession').text\n",
    "                prod_ver = prod_com.find('Gene-commentary_version').text\n",
    "                terms.append(prod_seq)\n",
    "            \n",
    "        # Lookup pubs\n",
    "        query = \" or \".join([\"'%s'\" % (i) for i in terms])\n",
    "        handle = Entrez.esearch(db=\"pubmed\",term=\"(\" + query + \") and '\"+ organism_name + \"'\")\n",
    "        search_results =Entrez.read(handle)\n",
    "\n",
    "        html += \"<tr>\"\n",
    "        \n",
    "        html += \"<td>\" + tag + \"</td>\"\n",
    "        html += \"<td>\" + locus + \"</td>\"\n",
    "        html += \"<td>\" + prot + \"</td>\"\n",
    "        html += \"<td>\" + prod_seq + \".\" + prod_ver + \"</td>\"\n",
    "        html += \"<td>\" + str(search_results['Count']) + \"</td>\" \n",
    "\n",
    "        html += \"</tr>\"\n",
    "    html += \"</table>\"\n",
    "\n",
    "    if len(genes) > max_show:\n",
    "        html += \"<p>And \" + str(len(genes) - max_show) + \" more... </p>\"\n",
    "        \n",
    "    return html\n",
    "\n",
    "\n",
    "def list_genes(accession, max_show = 100, look_gene_groups=False):\n",
    "    \"\"\"\n",
    "    For the given organism assembly build a table with first few genes.\n",
    "    \"\"\"\n",
    "    refseq_id, organism_name = get_asm_details(accession)\n",
    "\n",
    "    html = \"<h1>\" + organism_name + \"</h1>\"\n",
    "    \n",
    "    handle = Entrez.esearch(db=\"pubmed\",term=\"'\" + organism_name + \"'\")\n",
    "    search_results =Entrez.read(handle)\n",
    "\n",
    "    html += \"<p>\" + str(search_results['Count']) + \" pubmeds found</p>\" \n",
    "\n",
    "    if look_gene_groups:\n",
    "        genes = list_gene_ids_alt(refseq_id)\n",
    "    else:\n",
    "        nuccore_id = get_nuccore_id(refseq_id)\n",
    "        genes = list_gene_ids(nuccore_id)\n",
    "        \n",
    "        \n",
    "    html += \"<h3> Genes: </h3>\"\n",
    "    \n",
    "    html += build_gene_table(genes, max_show, organism_name)    \n",
    "\n",
    "    display(HTML(html))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "hide_input": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\n",
      "<?xml version=\"1.0\" ?>\n",
      "<!DOCTYPE Entrezgene-Set PUBLIC \"-//NLM//DTD NCBI-Entrezgene, 21st January 2005//EN\" \"https://www.ncbi.nlm.nih.gov/data_specs/dtd/NCBI_Entrezgene.dtd\">\n",
      "<Entrezgene-Set>\n",
      "<Entrezgene>\n",
      "  <Entrezgene_track-info>\n",
      "    <Gene-track>\n",
      "      <Gene-track_geneid>16830705</Gene-track_geneid>\n",
      "      <Gene-track_create-date>\n",
      "        <Date>\n",
      "          <Date_std>\n",
      "            <Date-std>\n",
      "              <Date-std_year>2013</Date-std_year>\n",
      "              <Date-std_month>9</Date-std_month>\n",
      "              <Date-std_day>11</Date-std_day>\n",
      "              <Date-std_hour>18</Date-std_hour>\n",
      "              <Date-std_minute>20</Date-std_minute>\n",
      "              <Date-std_second>0</Date-std_second>\n",
      "            </Date-std>\n",
      "          </Date_std>\n",
      "        </Date>\n",
      "      </Gene-track_create-date>\n",
      "      <Gene-track_update-date>\n",
      "        <Date>\n",
      "          <Date_std>\n",
      "            <Date-std>\n",
      "              <Date-std_year>2016</Date-std_year>\n",
      "              <Date-std_month>4</Date-std_month>\n",
      "              <Date-std_day>14</Date-std_day>\n",
      "              <Date-std_hour>21</Date-std_hour>\n",
      "              <Date-std_minute>4</Date-std_minute>\n",
      "              <Date-std_second>48</Date-std_second>\n",
      "            </Date-std>\n",
      "          </Date_std>\n",
      "        </Date>\n",
      "      </Gene-track_update-date>\n",
      "    </Gene-track>\n",
      "  </Entrezgene_track-info>\n",
      "  <Entrezgene_type value=\"protein-coding\">6</Entrezgene_type>\n",
      "  <Entrezgene_source>\n",
      "    <BioSource>\n",
      "      <BioSource_genome value=\"chromosome\">21</BioSource_genome>\n",
      "      <BioSource_org>\n",
      "        <Org-ref>\n",
      "          <Org-ref_taxname>Staphylococcus aureus subsp. aureus NCTC 8325</Org-ref_taxname>\n",
      "          <Org-ref_db>\n",
      "            <Dbtag>\n",
      "              <Dbtag_db>taxon</Dbtag_db>\n",
      "              <Dbtag_tag>\n",
      "                <Object-id>\n",
      "                  <Object-id_id>93061</Object-id_id>\n",
      "                </Object-id>\n",
      "              </Dbtag_tag>\n",
      "            </Dbtag>\n",
      "          </Org-ref_db>\n",
      "          <Org-ref_orgname>\n",
      "            <OrgName>\n",
      "              <OrgName_name>\n",
      "                <OrgName_name_binomial>\n",
      "                  <BinomialOrgName>\n",
      "                    <BinomialOrgName_genus>Staphylococcus</BinomialOrgName_genus>\n",
      "                    <BinomialOrgName_species>aureus</BinomialOrgName_species>\n",
      "                    <BinomialOrgName_subspecies>subsp. aureus</BinomialOrgName_subspecies>\n",
      "                  </BinomialOrgName>\n",
      "                </OrgName_name_binomial>\n",
      "              </OrgName_name>\n",
      "              <OrgName_mod>\n",
      "                <OrgMod>\n",
      "                  <OrgMod_subtype value=\"strain\">2</OrgMod_subtype>\n",
      "                  <OrgMod_subname>NCTC 8325</OrgMod_subname>\n",
      "                </OrgMod>\n",
      "                <OrgMod>\n",
      "                  <OrgMod_subtype value=\"sub-species\">22</OrgMod_subtype>\n",
      "                  <OrgMod_subname>aureus</OrgMod_subname>\n",
      "                </OrgMod>\n",
      "              </OrgName_mod>\n",
      "              <OrgName_lineage>Bacteria; Firmicutes; Bacilli; Bacillales; Staphylococcaceae; Staphylococcus</OrgName_lineage>\n",
      "              <OrgName_gcode>11</OrgName_gcode>\n",
      "              <OrgName_div>BCT</OrgName_div>\n",
      "            </OrgName>\n",
      "          </Org-ref_orgname>\n",
      "        </Org-ref>\n",
      "      </BioSource_org>\n",
      "    </BioSource>\n",
      "  </Entrezgene_source>\n",
      "  <Entrezgene_gene>\n",
      "    <Gene-ref>\n",
      "      <Gene-ref_locus-tag>SAOUHSC_3042a</Gene-ref_locus-tag>\n",
      "    </Gene-ref>\n",
      "  </Entrezgene_gene>\n",
      "  <Entrezgene_prot>\n",
      "    <Prot-ref>\n",
      "      <Prot-ref_name>\n",
      "        <Prot-ref_name_E>SAOUHSC_13820</Prot-ref_name_E>\n",
      "      </Prot-ref_name>\n",
      "    </Prot-ref>\n",
      "  </Entrezgene_prot>\n",
      "  <Entrezgene_gene-source>\n",
      "    <Gene-source>\n",
      "      <Gene-source_src>Entrez Genomes</Gene-source_src>\n",
      "      <Gene-source_src-int>88193823</Gene-source_src-int>\n",
      "      <Gene-source_src-str1>NC_007795</Gene-source_src-str1>\n",
      "      <Gene-source_src-str2>SAOUHSC_3042a</Gene-source_src-str2>\n",
      "    </Gene-source>\n",
      "  </Entrezgene_gene-source>\n",
      "  <Entrezgene_locus>\n",
      "    <Gene-commentary>\n",
      "      <Gene-commentary_type value=\"genomic\">1</Gene-commentary_type>\n",
      "      <Gene-commentary_accession>NC_007795</Gene-commentary_accession>\n",
      "      <Gene-commentary_version>1</Gene-commentary_version>\n",
      "      <Gene-commentary_seqs>\n",
      "        <Seq-loc>\n",
      "          <Seq-loc_int>\n",
      "            <Seq-interval>\n",
      "              <Seq-interval_from>2812462</Seq-interval_from>\n",
      "              <Seq-interval_to>2812653</Seq-interval_to>\n",
      "              <Seq-interval_strand>\n",
      "                <Na-strand value=\"plus\"/>\n",
      "              </Seq-interval_strand>\n",
      "              <Seq-interval_id>\n",
      "                <Seq-id>\n",
      "                  <Seq-id_gi>88193823</Seq-id_gi>\n",
      "                </Seq-id>\n",
      "              </Seq-interval_id>\n",
      "            </Seq-interval>\n",
      "          </Seq-loc_int>\n",
      "        </Seq-loc>\n",
      "      </Gene-commentary_seqs>\n",
      "      <Gene-commentary_products>\n",
      "        <Gene-commentary>\n",
      "          <Gene-commentary_type value=\"peptide\">8</Gene-commentary_type>\n",
      "          <Gene-commentary_label>SAOUHSC_13820</Gene-commentary_label>\n",
      "          <Gene-commentary_accession>YP_008530245</Gene-commentary_accession>\n",
      "          <Gene-commentary_version>1</Gene-commentary_version>\n",
      "          <Gene-commentary_genomic-coords>\n",
      "            <Seq-loc>\n",
      "              <Seq-loc_int>\n",
      "                <Seq-interval>\n",
      "                  <Seq-interval_from>2812462</Seq-interval_from>\n",
      "                  <Seq-interval_to>2812653</Seq-interval_to>\n",
      "                  <Seq-interval_strand>\n",
      "                    <Na-strand value=\"plus\"/>\n",
      "                  </Seq-interval_strand>\n",
      "                  <Seq-interval_id>\n",
      "                    <Seq-id>\n",
      "                      <Seq-id_gi>88193823</Seq-id_gi>\n",
      "                    </Seq-id>\n",
      "                  </Seq-interval_id>\n",
      "                </Seq-interval>\n",
      "              </Seq-loc_int>\n",
      "            </Seq-loc>\n",
      "          </Gene-commentary_genomic-coords>\n",
      "          <Gene-commentary_seqs>\n",
      "            <Seq-loc>\n",
      "              <Seq-loc_whole>\n",
      "                <Seq-id>\n",
      "                  <Seq-id_gi>542123278</Seq-id_gi>\n",
      "                </Seq-id>\n",
      "              </Seq-loc_whole>\n",
      "            </Seq-loc>\n",
      "          </Gene-commentary_seqs>\n",
      "        </Gene-commentary>\n",
      "      </Gene-commentary_products>\n",
      "    </Gene-commentary>\n",
      "  </Entrezgene_locus>\n",
      "  <Entrezgene_comments>\n",
      "    <Gene-commentary>\n",
      "      <Gene-commentary_type value=\"comment\">254</Gene-commentary_type>\n",
      "      <Gene-commentary_heading>NCBI Reference Sequences (RefSeq)</Gene-commentary_heading>\n",
      "      <Gene-commentary_comment>\n",
      "        <Gene-commentary>\n",
      "          <Gene-commentary_type value=\"comment\">254</Gene-commentary_type>\n",
      "          <Gene-commentary_heading>Genome Annotation</Gene-commentary_heading>\n",
      "          <Gene-commentary_products>\n",
      "            <Gene-commentary>\n",
      "              <Gene-commentary_type value=\"genomic\">1</Gene-commentary_type>\n",
      "              <Gene-commentary_heading>Reference assembly</Gene-commentary_heading>\n",
      "              <Gene-commentary_accession>NC_007795</Gene-commentary_accession>\n",
      "              <Gene-commentary_version>1</Gene-commentary_version>\n",
      "              <Gene-commentary_source>\n",
      "                <Other-source>\n",
      "                  <Other-source_src>\n",
      "                    <Dbtag>\n",
      "                      <Dbtag_db>Nucleotide</Dbtag_db>\n",
      "                      <Dbtag_tag>\n",
      "                        <Object-id>\n",
      "                          <Object-id_id>88193823</Object-id_id>\n",
      "                        </Object-id>\n",
      "                      </Dbtag_tag>\n",
      "                    </Dbtag>\n",
      "                  </Other-source_src>\n",
      "                </Other-source>\n",
      "              </Gene-commentary_source>\n",
      "              <Gene-commentary_seqs>\n",
      "                <Seq-loc>\n",
      "                  <Seq-loc_int>\n",
      "                    <Seq-interval>\n",
      "                      <Seq-interval_from>2812462</Seq-interval_from>\n",
      "                      <Seq-interval_to>2812653</Seq-interval_to>\n",
      "                      <Seq-interval_strand>\n",
      "                        <Na-strand value=\"plus\"/>\n",
      "                      </Seq-interval_strand>\n",
      "                      <Seq-interval_id>\n",
      "                        <Seq-id>\n",
      "                          <Seq-id_gi>88193823</Seq-id_gi>\n",
      "                        </Seq-id>\n",
      "                      </Seq-interval_id>\n",
      "                    </Seq-interval>\n",
      "                  </Seq-loc_int>\n",
      "                </Seq-loc>\n",
      "              </Gene-commentary_seqs>\n",
      "              <Gene-commentary_products>\n",
      "                <Gene-commentary>\n",
      "                  <Gene-commentary_type value=\"peptide\">8</Gene-commentary_type>\n",
      "                  <Gene-commentary_heading>Product</Gene-commentary_heading>\n",
      "                  <Gene-commentary_accession>YP_008530245</Gene-commentary_accession>\n",
      "                  <Gene-commentary_version>1</Gene-commentary_version>\n",
      "                  <Gene-commentary_source>\n",
      "                    <Other-source>\n",
      "                      <Other-source_src>\n",
      "                        <Dbtag>\n",
      "                          <Dbtag_db>Protein</Dbtag_db>\n",
      "                          <Dbtag_tag>\n",
      "                            <Object-id>\n",
      "                              <Object-id_id>542123278</Object-id_id>\n",
      "                            </Object-id>\n",
      "                          </Dbtag_tag>\n",
      "                        </Dbtag>\n",
      "                      </Other-source_src>\n",
      "                      <Other-source_anchor>YP_008530245</Other-source_anchor>\n",
      "                      <Other-source_post-text>SAOUHSC_13820 [Staphylococcus aureus subsp. aureus NCTC 8325]</Other-source_post-text>\n",
      "                    </Other-source>\n",
      "                  </Gene-commentary_source>\n",
      "                  <Gene-commentary_seqs>\n",
      "                    <Seq-loc>\n",
      "                      <Seq-loc_whole>\n",
      "                        <Seq-id>\n",
      "                          <Seq-id_gi>542123278</Seq-id_gi>\n",
      "                        </Seq-id>\n",
      "                      </Seq-loc_whole>\n",
      "                    </Seq-loc>\n",
      "                  </Gene-commentary_seqs>\n",
      "                  <Gene-commentary_comment>\n",
      "                    <Gene-commentary>\n",
      "                      <Gene-commentary_type value=\"comment\">254</Gene-commentary_type>\n",
      "                      <Gene-commentary_label>RefSeq Status</Gene-commentary_label>\n",
      "                      <Gene-commentary_text>REVIEWED</Gene-commentary_text>\n",
      "                    </Gene-commentary>\n",
      "                  </Gene-commentary_comment>\n",
      "                </Gene-commentary>\n",
      "              </Gene-commentary_products>\n",
      "            </Gene-commentary>\n",
      "          </Gene-commentary_products>\n",
      "        </Gene-commentary>\n",
      "      </Gene-commentary_comment>\n",
      "    </Gene-commentary>\n",
      "    <Gene-commentary>\n",
      "      <Gene-commentary_type value=\"comment\">254</Gene-commentary_type>\n",
      "      <Gene-commentary_heading>RefSeq Status</Gene-commentary_heading>\n",
      "      <Gene-commentary_label>REVIEWED</Gene-commentary_label>\n",
      "    </Gene-commentary>\n",
      "  </Entrezgene_comments>\n",
      "  <Entrezgene_unique-keys></Entrezgene_unique-keys>\n",
      "  <Entrezgene_xtra-iq></Entrezgene_xtra-iq>\n",
      "  <Entrezgene_non-unique-keys>\n",
      "    <Dbtag>\n",
      "      <Dbtag_db>ID</Dbtag_db>\n",
      "      <Dbtag_tag>\n",
      "        <Object-id>\n",
      "          <Object-id_id>88193823</Object-id_id>\n",
      "        </Object-id>\n",
      "      </Dbtag_tag>\n",
      "    </Dbtag>\n",
      "  </Entrezgene_non-unique-keys>\n",
      "</Entrezgene>\n",
      "</Entrezgene-Set>\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<h1>Staphylococcus aureus</h1><p>24 pubmeds found</p><h3> Genes: </h3>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 40 ms, sys: 4 ms, total: 44 ms\n",
      "Wall time: 1.14 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "list_genes('GCF_000013425.1', 5, False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\n",
      "<?xml version=\"1.0\" ?>\n",
      "<!DOCTYPE Entrezgene-Set PUBLIC \"-//NLM//DTD NCBI-Entrezgene, 21st January 2005//EN\" \"https://www.ncbi.nlm.nih.gov/data_specs/dtd/NCBI_Entrezgene.dtd\">\n",
      "<Entrezgene-Set>\n",
      "<Entrezgene>\n",
      "  <Entrezgene_track-info>\n",
      "    <Gene-track>\n",
      "      <Gene-track_geneid>31215633</Gene-track_geneid>\n",
      "      <Gene-track_create-date>\n",
      "        <Date>\n",
      "          <Date_std>\n",
      "            <Date-std>\n",
      "              <Date-std_year>2017</Date-std_year>\n",
      "              <Date-std_month>3</Date-std_month>\n",
      "              <Date-std_day>5</Date-std_day>\n",
      "              <Date-std_hour>13</Date-std_hour>\n",
      "              <Date-std_minute>33</Date-std_minute>\n",
      "              <Date-std_second>0</Date-std_second>\n",
      "            </Date-std>\n",
      "          </Date_std>\n",
      "        </Date>\n",
      "      </Gene-track_create-date>\n",
      "      <Gene-track_update-date>\n",
      "        <Date>\n",
      "          <Date_std>\n",
      "            <Date-std>\n",
      "              <Date-std_year>2017</Date-std_year>\n",
      "              <Date-std_month>9</Date-std_month>\n",
      "              <Date-std_day>8</Date-std_day>\n",
      "              <Date-std_hour>14</Date-std_hour>\n",
      "              <Date-std_minute>32</Date-std_minute>\n",
      "              <Date-std_second>0</Date-std_second>\n",
      "            </Date-std>\n",
      "          </Date_std>\n",
      "        </Date>\n",
      "      </Gene-track_update-date>\n",
      "    </Gene-track>\n",
      "  </Entrezgene_track-info>\n",
      "  <Entrezgene_type value=\"protein-coding\">6</Entrezgene_type>\n",
      "  <Entrezgene_source>\n",
      "    <BioSource>\n",
      "      <BioSource_genome value=\"chromosome\">21</BioSource_genome>\n",
      "      <BioSource_org>\n",
      "        <Org-ref>\n",
      "          <Org-ref_taxname>Staphylococcus aureus</Org-ref_taxname>\n",
      "          <Org-ref_db>\n",
      "            <Dbtag>\n",
      "              <Dbtag_db>taxon</Dbtag_db>\n",
      "              <Dbtag_tag>\n",
      "                <Object-id>\n",
      "                  <Object-id_id>1280</Object-id_id>\n",
      "                </Object-id>\n",
      "              </Dbtag_tag>\n",
      "            </Dbtag>\n",
      "          </Org-ref_db>\n",
      "          <Org-ref_syn>\n",
      "            <Org-ref_syn_E>Micrococcus aureus</Org-ref_syn_E>\n",
      "            <Org-ref_syn_E>Micrococcus pyogenes</Org-ref_syn_E>\n",
      "            <Org-ref_syn_E>Staphlococcus pyogenes citreus</Org-ref_syn_E>\n",
      "            <Org-ref_syn_E>Staphylococcus pyogenes aureus</Org-ref_syn_E>\n",
      "          </Org-ref_syn>\n",
      "          <Org-ref_orgname>\n",
      "            <OrgName>\n",
      "              <OrgName_name>\n",
      "                <OrgName_name_binomial>\n",
      "                  <BinomialOrgName>\n",
      "                    <BinomialOrgName_genus>Staphylococcus</BinomialOrgName_genus>\n",
      "                    <BinomialOrgName_species>aureus</BinomialOrgName_species>\n",
      "                  </BinomialOrgName>\n",
      "                </OrgName_name_binomial>\n",
      "              </OrgName_name>\n",
      "              <OrgName_mod>\n",
      "                <OrgMod>\n",
      "                  <OrgMod_subtype value=\"strain\">2</OrgMod_subtype>\n",
      "                  <OrgMod_subname>MS4</OrgMod_subname>\n",
      "                </OrgMod>\n",
      "                <OrgMod>\n",
      "                  <OrgMod_subtype value=\"nat-host\">21</OrgMod_subtype>\n",
      "                  <OrgMod_subname>Homo sapiens</OrgMod_subname>\n",
      "                </OrgMod>\n",
      "              </OrgName_mod>\n",
      "              <OrgName_lineage>Bacteria; Firmicutes; Bacilli; Bacillales; Staphylococcaceae; Staphylococcus</OrgName_lineage>\n",
      "              <OrgName_gcode>11</OrgName_gcode>\n",
      "              <OrgName_div>BCT</OrgName_div>\n",
      "            </OrgName>\n",
      "          </Org-ref_orgname>\n",
      "        </Org-ref>\n",
      "      </BioSource_org>\n",
      "      <BioSource_subtype>\n",
      "        <SubSource>\n",
      "          <SubSource_subtype value=\"country\">23</SubSource_subtype>\n",
      "          <SubSource_name>China: Shenzhen</SubSource_name>\n",
      "        </SubSource>\n",
      "        <SubSource>\n",
      "          <SubSource_subtype value=\"isolation-source\">28</SubSource_subtype>\n",
      "          <SubSource_name>the wound exudate of bone fracture</SubSource_name>\n",
      "        </SubSource>\n",
      "        <SubSource>\n",
      "          <SubSource_subtype value=\"lat-lon\">29</SubSource_subtype>\n",
      "          <SubSource_name>22.5313 N 113.924 E</SubSource_name>\n",
      "        </SubSource>\n",
      "        <SubSource>\n",
      "          <SubSource_subtype value=\"collection-date\">30</SubSource_subtype>\n",
      "          <SubSource_name>05-Jul-2012</SubSource_name>\n",
      "        </SubSource>\n",
      "        <SubSource>\n",
      "          <SubSource_subtype value=\"collected-by\">31</SubSource_subtype>\n",
      "          <SubSource_name>Deng Qiwen</SubSource_name>\n",
      "        </SubSource>\n",
      "      </BioSource_subtype>\n",
      "    </BioSource>\n",
      "  </Entrezgene_source>\n",
      "  <Entrezgene_gene>\n",
      "    <Gene-ref>\n",
      "      <Gene-ref_locus-tag>NI36_RS14355</Gene-ref_locus-tag>\n",
      "    </Gene-ref>\n",
      "  </Entrezgene_gene>\n",
      "  <Entrezgene_prot>\n",
      "    <Prot-ref>\n",
      "      <Prot-ref_name>\n",
      "        <Prot-ref_name_E>hypothetical protein</Prot-ref_name_E>\n",
      "      </Prot-ref_name>\n",
      "    </Prot-ref>\n",
      "  </Entrezgene_prot>\n",
      "  <Entrezgene_gene-source>\n",
      "    <Gene-source>\n",
      "      <Gene-source_src>Entrez Genomes</Gene-source_src>\n",
      "      <Gene-source_src-int>959863495</Gene-source_src-int>\n",
      "      <Gene-source_src-str1>NZ_CP009828</Gene-source_src-str1>\n",
      "      <Gene-source_src-str2>NI36_RS14355</Gene-source_src-str2>\n",
      "    </Gene-source>\n",
      "  </Entrezgene_gene-source>\n",
      "  <Entrezgene_locus>\n",
      "    <Gene-commentary>\n",
      "      <Gene-commentary_type value=\"genomic\">1</Gene-commentary_type>\n",
      "      <Gene-commentary_accession>NZ_CP009828</Gene-commentary_accession>\n",
      "      <Gene-commentary_version>1</Gene-commentary_version>\n",
      "      <Gene-commentary_seqs>\n",
      "        <Seq-loc>\n",
      "          <Seq-loc_int>\n",
      "            <Seq-interval>\n",
      "              <Seq-interval_from>2703290</Seq-interval_from>\n",
      "              <Seq-interval_to>2703367</Seq-interval_to>\n",
      "              <Seq-interval_strand>\n",
      "                <Na-strand value=\"plus\"/>\n",
      "              </Seq-interval_strand>\n",
      "              <Seq-interval_id>\n",
      "                <Seq-id>\n",
      "                  <Seq-id_gi>959863495</Seq-id_gi>\n",
      "                </Seq-id>\n",
      "              </Seq-interval_id>\n",
      "            </Seq-interval>\n",
      "          </Seq-loc_int>\n",
      "        </Seq-loc>\n",
      "      </Gene-commentary_seqs>\n",
      "      <Gene-commentary_products>\n",
      "        <Gene-commentary>\n",
      "          <Gene-commentary_type value=\"peptide\">8</Gene-commentary_type>\n",
      "          <Gene-commentary_label>hypothetical protein</Gene-commentary_label>\n",
      "          <Gene-commentary_accession>WP_078061161</Gene-commentary_accession>\n",
      "          <Gene-commentary_version>1</Gene-commentary_version>\n",
      "          <Gene-commentary_genomic-coords>\n",
      "            <Seq-loc>\n",
      "              <Seq-loc_int>\n",
      "                <Seq-interval>\n",
      "                  <Seq-interval_from>2703290</Seq-interval_from>\n",
      "                  <Seq-interval_to>2703367</Seq-interval_to>\n",
      "                  <Seq-interval_strand>\n",
      "                    <Na-strand value=\"plus\"/>\n",
      "                  </Seq-interval_strand>\n",
      "                  <Seq-interval_id>\n",
      "                    <Seq-id>\n",
      "                      <Seq-id_gi>959863495</Seq-id_gi>\n",
      "                    </Seq-id>\n",
      "                  </Seq-interval_id>\n",
      "                </Seq-interval>\n",
      "              </Seq-loc_int>\n",
      "            </Seq-loc>\n",
      "          </Gene-commentary_genomic-coords>\n",
      "          <Gene-commentary_seqs>\n",
      "            <Seq-loc>\n",
      "              <Seq-loc_whole>\n",
      "                <Seq-id>\n",
      "                  <Seq-id_gi>1151011228</Seq-id_gi>\n",
      "                </Seq-id>\n",
      "              </Seq-loc_whole>\n",
      "            </Seq-loc>\n",
      "          </Gene-commentary_seqs>\n",
      "        </Gene-commentary>\n",
      "      </Gene-commentary_products>\n",
      "    </Gene-commentary>\n",
      "  </Entrezgene_locus>\n",
      "  <Entrezgene_comments>\n",
      "    <Gene-commentary>\n",
      "      <Gene-commentary_type value=\"comment\">254</Gene-commentary_type>\n",
      "      <Gene-commentary_heading>NCBI Reference Sequences (RefSeq)</Gene-commentary_heading>\n",
      "      <Gene-commentary_comment>\n",
      "        <Gene-commentary>\n",
      "          <Gene-commentary_type value=\"comment\">254</Gene-commentary_type>\n",
      "          <Gene-commentary_heading>Genome Annotation</Gene-commentary_heading>\n",
      "          <Gene-commentary_products>\n",
      "            <Gene-commentary>\n",
      "              <Gene-commentary_type value=\"genomic\">1</Gene-commentary_type>\n",
      "              <Gene-commentary_heading>Reference assembly</Gene-commentary_heading>\n",
      "              <Gene-commentary_accession>NZ_CP009828</Gene-commentary_accession>\n",
      "              <Gene-commentary_version>1</Gene-commentary_version>\n",
      "              <Gene-commentary_source>\n",
      "                <Other-source>\n",
      "                  <Other-source_src>\n",
      "                    <Dbtag>\n",
      "                      <Dbtag_db>Nucleotide</Dbtag_db>\n",
      "                      <Dbtag_tag>\n",
      "                        <Object-id>\n",
      "                          <Object-id_id>959863495</Object-id_id>\n",
      "                        </Object-id>\n",
      "                      </Dbtag_tag>\n",
      "                    </Dbtag>\n",
      "                  </Other-source_src>\n",
      "                </Other-source>\n",
      "              </Gene-commentary_source>\n",
      "              <Gene-commentary_seqs>\n",
      "                <Seq-loc>\n",
      "                  <Seq-loc_int>\n",
      "                    <Seq-interval>\n",
      "                      <Seq-interval_from>2703290</Seq-interval_from>\n",
      "                      <Seq-interval_to>2703367</Seq-interval_to>\n",
      "                      <Seq-interval_strand>\n",
      "                        <Na-strand value=\"plus\"/>\n",
      "                      </Seq-interval_strand>\n",
      "                      <Seq-interval_id>\n",
      "                        <Seq-id>\n",
      "                          <Seq-id_gi>959863495</Seq-id_gi>\n",
      "                        </Seq-id>\n",
      "                      </Seq-interval_id>\n",
      "                    </Seq-interval>\n",
      "                  </Seq-loc_int>\n",
      "                </Seq-loc>\n",
      "              </Gene-commentary_seqs>\n",
      "              <Gene-commentary_products>\n",
      "                <Gene-commentary>\n",
      "                  <Gene-commentary_type value=\"peptide\">8</Gene-commentary_type>\n",
      "                  <Gene-commentary_heading>Product</Gene-commentary_heading>\n",
      "                  <Gene-commentary_accession>WP_078061161</Gene-commentary_accession>\n",
      "                  <Gene-commentary_version>1</Gene-commentary_version>\n",
      "                  <Gene-commentary_source>\n",
      "                    <Other-source>\n",
      "                      <Other-source_src>\n",
      "                        <Dbtag>\n",
      "                          <Dbtag_db>Protein</Dbtag_db>\n",
      "                          <Dbtag_tag>\n",
      "                            <Object-id>\n",
      "                              <Object-id_id>1151011228</Object-id_id>\n",
      "                            </Object-id>\n",
      "                          </Dbtag_tag>\n",
      "                        </Dbtag>\n",
      "                      </Other-source_src>\n",
      "                      <Other-source_anchor>WP_078061161</Other-source_anchor>\n",
      "                      <Other-source_post-text>hypothetical protein NI36_RS14355 [Staphylococcus aureus]</Other-source_post-text>\n",
      "                    </Other-source>\n",
      "                  </Gene-commentary_source>\n",
      "                  <Gene-commentary_seqs>\n",
      "                    <Seq-loc>\n",
      "                      <Seq-loc_whole>\n",
      "                        <Seq-id>\n",
      "                          <Seq-id_gi>1151011228</Seq-id_gi>\n",
      "                        </Seq-id>\n",
      "                      </Seq-loc_whole>\n",
      "                    </Seq-loc>\n",
      "                  </Gene-commentary_seqs>\n",
      "                  <Gene-commentary_comment>\n",
      "                    <Gene-commentary>\n",
      "                      <Gene-commentary_type value=\"other\">255</Gene-commentary_type>\n",
      "                      <Gene-commentary_heading>Conserved Domains</Gene-commentary_heading>\n",
      "                      <Gene-commentary_source>\n",
      "                        <Other-source>\n",
      "                          <Other-source_pre-text>(1)</Other-source_pre-text>\n",
      "                          <Other-source_anchor>summary</Other-source_anchor>\n",
      "                        </Other-source>\n",
      "                      </Gene-commentary_source>\n",
      "                      <Gene-commentary_comment>\n",
      "                        <Gene-commentary>\n",
      "                          <Gene-commentary_type value=\"other\">255</Gene-commentary_type>\n",
      "                          <Gene-commentary_source>\n",
      "                            <Other-source>\n",
      "                              <Other-source_src>\n",
      "                                <Dbtag>\n",
      "                                  <Dbtag_db>CDD</Dbtag_db>\n",
      "                                  <Dbtag_tag>\n",
      "                                    <Object-id>\n",
      "                                      <Object-id_id>314300</Object-id_id>\n",
      "                                    </Object-id>\n",
      "                                  </Dbtag_tag>\n",
      "                                </Dbtag>\n",
      "                              </Other-source_src>\n",
      "                              <Other-source_anchor>pfam11323: DUF3125; Protein of unknown function (DUF3125)</Other-source_anchor>\n",
      "                            </Other-source>\n",
      "                          </Gene-commentary_source>\n",
      "                          <Gene-commentary_comment>\n",
      "                            <Gene-commentary>\n",
      "                              <Gene-commentary_type value=\"other\">255</Gene-commentary_type>\n",
      "                              <Gene-commentary_text>Location: 1 - 24</Gene-commentary_text>\n",
      "                            </Gene-commentary>\n",
      "                          </Gene-commentary_comment>\n",
      "                        </Gene-commentary>\n",
      "                      </Gene-commentary_comment>\n",
      "                    </Gene-commentary>\n",
      "                  </Gene-commentary_comment>\n",
      "                </Gene-commentary>\n",
      "              </Gene-commentary_products>\n",
      "            </Gene-commentary>\n",
      "          </Gene-commentary_products>\n",
      "        </Gene-commentary>\n",
      "      </Gene-commentary_comment>\n",
      "    </Gene-commentary>\n",
      "    <Gene-commentary>\n",
      "      <Gene-commentary_type value=\"comment\">254</Gene-commentary_type>\n",
      "      <Gene-commentary_heading>RefSeq Status</Gene-commentary_heading>\n",
      "      <Gene-commentary_label></Gene-commentary_label>\n",
      "    </Gene-commentary>\n",
      "  </Entrezgene_comments>\n",
      "  <Entrezgene_unique-keys></Entrezgene_unique-keys>\n",
      "  <Entrezgene_xtra-iq></Entrezgene_xtra-iq>\n",
      "  <Entrezgene_non-unique-keys>\n",
      "    <Dbtag>\n",
      "      <Dbtag_db>ID</Dbtag_db>\n",
      "      <Dbtag_tag>\n",
      "        <Object-id>\n",
      "          <Object-id_id>959863495</Object-id_id>\n",
      "        </Object-id>\n",
      "      </Dbtag_tag>\n",
      "    </Dbtag>\n",
      "  </Entrezgene_non-unique-keys>\n",
      "</Entrezgene>\n",
      "</Entrezgene-Set>\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<h1>Staphylococcus aureus</h1><p>24 pubmeds found</p><h3> Genes: </h3>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "list_genes('GCF_000013465.1', 10, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h1>Staphylococcus aureus</h1><p>24 pubmeds found</p><h3> Genes: </h3><table><tr><th> Gene Tag </th><th> Gene Locus </th><th> Protein </th><th> Protein sequence </th><th> Number of Pubmed publications </th></tr><tr><td>NI36_RS14355</td><td>n/a</td><td>hypothetical protein</td><td>WP_078061161.1</td><td>0</td></tr><tr><td>NI36_RS14350</td><td>n/a</td><td>hypothetical protein</td><td>WP_071621415.1</td><td>0</td></tr><tr><td>NI36_RS14345</td><td>n/a</td><td>n/a</td><td>n/a.n/a</td><td>0</td></tr><tr><td>NI36_RS14340</td><td>n/a</td><td>n/a</td><td>n/a.n/a</td><td>0</td></tr><tr><td>NI36_RS14335</td><td>n/a</td><td>n/a</td><td>n/a.n/a</td><td>0</td></tr><tr><td>NI36_RS14330</td><td>n/a</td><td>n/a</td><td>n/a.n/a</td><td>0</td></tr><tr><td>NI36_RS14325</td><td>n/a</td><td>hypothetical protein</td><td>WP_078065189.1</td><td>0</td></tr><tr><td>NI36_RS14320</td><td>n/a</td><td>hypothetical protein</td><td>WP_001228565.1</td><td>0</td></tr><tr><td>NI36_RS14315</td><td>n/a</td><td>DUF2648 domain-containing protein</td><td>WP_000731972.1</td><td>0</td></tr><tr><td>NI36_RS14310</td><td>n/a</td><td>hypothetical protein</td><td>WP_001803564.1</td><td>0</td></tr><tr><td>NI36_RS14305</td><td>n/a</td><td>hypothetical protein</td><td>WP_078065156.1</td><td>0</td></tr><tr><td>NI36_RS14300</td><td>n/a</td><td>n/a</td><td>n/a.n/a</td><td>0</td></tr><tr><td>NI36_RS14295</td><td>n/a</td><td>n/a</td><td>n/a.n/a</td><td>0</td></tr><tr><td>NI36_RS14290</td><td>n/a</td><td>hypothetical protein</td><td>WP_071665647.1</td><td>0</td></tr><tr><td>NI36_RS14285</td><td>n/a</td><td>hypothetical protein</td><td>WP_001621913.1</td><td>0</td></tr><tr><td>NI36_RS14280</td><td>n/a</td><td>hypothetical protein</td><td>WP_078065154.1</td><td>0</td></tr><tr><td>NI36_RS14275</td><td>n/a</td><td>n/a</td><td>n/a.n/a</td><td>0</td></tr><tr><td>NI36_RS14270</td><td>n/a</td><td>hypothetical protein</td><td>WP_001792003.1</td><td>0</td></tr><tr><td>NI36_RS14265</td><td>n/a</td><td>hypothetical protein</td><td>WP_071665669.1</td><td>0</td></tr><tr><td>NI36_RS14260</td><td>n/a</td><td>hypothetical protein</td><td>WP_071665646.1</td><td>0</td></tr><tr><td>NI36_RS14255</td><td>n/a</td><td>hypothetical protein</td><td>WP_031845057.1</td><td>0</td></tr><tr><td>NI36_RS14250</td><td>n/a</td><td>n/a</td><td>n/a.n/a</td><td>0</td></tr><tr><td>NI36_RS14245</td><td>n/a</td><td>hypothetical protein</td><td>WP_001789152.1</td><td>0</td></tr><tr><td>NI36_RS14240</td><td>n/a</td><td>hypothetical protein</td><td>WP_071665644.1</td><td>0</td></tr><tr><td>NI36_RS14235</td><td>n/a</td><td>hypothetical protein</td><td>WP_071665643.1</td><td>0</td></tr><tr><td>NI36_RS14230</td><td>n/a</td><td>hypothetical protein</td><td>WP_031927355.1</td><td>0</td></tr><tr><td>NI36_RS14225</td><td>n/a</td><td>n/a</td><td>n/a.n/a</td><td>0</td></tr><tr><td>NI36_RS14220</td><td>n/a</td><td>hypothetical protein</td><td>WP_071665642.1</td><td>0</td></tr><tr><td>NI36_RS14215</td><td>n/a</td><td>hypothetical protein</td><td>n/a.n/a</td><td>0</td></tr><tr><td>NI36_RS14210</td><td>n/a</td><td>n/a</td><td>n/a.n/a</td><td>0</td></tr><tr><td>NI36_RS14205</td><td>n/a</td><td>n/a</td><td>n/a.n/a</td><td>0</td></tr><tr><td>NI36_RS14200</td><td>n/a</td><td>n/a</td><td>n/a.n/a</td><td>0</td></tr><tr><td>NI36_RS14195</td><td>n/a</td><td>n/a</td><td>n/a.n/a</td><td>0</td></tr><tr><td>NI36_RS14190</td><td>n/a</td><td>n/a</td><td>n/a.n/a</td><td>0</td></tr><tr><td>NI36_RS14185</td><td>n/a</td><td>hypothetical protein</td><td>WP_000312015.1</td><td>0</td></tr><tr><td>NI36_RS14180</td><td>n/a</td><td>hypothetical protein</td><td>WP_071665667.1</td><td>0</td></tr><tr><td>NI36_RS14175</td><td>n/a</td><td>hypothetical protein</td><td>WP_071665666.1</td><td>0</td></tr><tr><td>NI36_RS14170</td><td>n/a</td><td>hypothetical protein</td><td>WP_071665640.1</td><td>0</td></tr><tr><td>NI36_RS14165</td><td>n/a</td><td>n/a</td><td>n/a.n/a</td><td>0</td></tr><tr><td>NI36_RS14160</td><td>n/a</td><td>hypothetical protein</td><td>WP_071665639.1</td><td>0</td></tr><tr><td>NI36_RS14155</td><td>n/a</td><td>hypothetical protein</td><td>WP_071665638.1</td><td>0</td></tr><tr><td>NI36_RS14150</td><td>n/a</td><td>hypothetical protein</td><td>WP_071665636.1</td><td>0</td></tr><tr><td>NI36_RS14145</td><td>n/a</td><td>n/a</td><td>n/a.n/a</td><td>0</td></tr><tr><td>NI36_RS14140</td><td>n/a</td><td>hypothetical protein</td><td>WP_071665635.1</td><td>0</td></tr><tr><td>NI36_RS14135</td><td>n/a</td><td>n/a</td><td>n/a.n/a</td><td>0</td></tr><tr><td>NI36_RS14130</td><td>n/a</td><td>hypothetical protein</td><td>WP_001791824.1</td><td>0</td></tr><tr><td>NI36_RS14125</td><td>n/a</td><td>hypothetical protein</td><td>WP_001791825.1</td><td>0</td></tr><tr><td>NI36_RS14120</td><td>n/a</td><td>hypothetical protein</td><td>WP_001795394.1</td><td>0</td></tr><tr><td>NI36_RS14115</td><td>n/a</td><td>n/a</td><td>n/a.n/a</td><td>0</td></tr><tr><td>NI36_RS14110</td><td>n/a</td><td>n/a</td><td>n/a.n/a</td><td>0</td></tr></table><p>And 6236 more... </p>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "list_genes('GCF_000013425.1', 50, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
